### PR TITLE
Martin/jib 95 show dotted outline for sending a new message or reply

### PR DESCRIPTION
**Pull request by MartinJibber**
https://user-images.githubusercontent.com/88675954/137242475-174ba4f0-db45-4ec1-a360-db632f24497b.mov

### CODE REVIEW

File: Shared/Extensions/UI Extensions/UICollectionView+Extensions.swift [hunk 1]
```diff
@@ -10,8 +10,17 @@ import Foundation
...
     func getCentermostVisibleIndex() -> IndexPath? {
+        if let closestCell = self.getCentermostVisibleCell() {
+            return self.indexPath(for: closestCell)
+        }
+
+        return nil
+    }
+
+    /// Gets the index of the visible cell that is closest to the center of this collection view.
+    func getCentermostVisibleCell() -> UICollectionViewCell? {
```
Comment (MartinJibber) [inline Shared/Extensions/UI Extensions/UICollectionView+Extensions.swift:23]: Adding a convenience function to get the centermost cell, not just the index path. Note that the index path version uses this new function so we don't duplicate logic.

File: iOS/Flows/Archives/Conversation/ConversationContentView.swift [hunk 2]
```diff
@@ -44,11 +44,6 @@ class ConversationContentView: View {
     }
 
     func configure(with item: Conversation) {
-
-        defer {
-            self.currentItem = item
-        }
-
```
Comment (MartinJibber) [inline iOS/Flows/Archives/Conversation/ConversationContentView.swift:51]: The Xcode 13 simulators are all kinds of messed up so I'm using the beta for now. Unfortunately the beta doesn't know how to handle this defer statement so I'm just putting it's body at the end of the function.

File: iOS/Flows/Conversation/ConversationCollectionViewLayout.swift [hunk 3]
```diff
@@ -35,14 +35,15 @@ class ConversationCollectionViewLayout: UICollectionViewFlowLayout {
                                                    bottom: 0,
                                                    right: collectionView.width * 0.1)
 
-        let itemHeight: CGFloat = 200
+        let itemHeight: CGFloat = collectionView.height * 0.27
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/ConversationCollectionViewLayout.swift:38]: Making the height proportional to the device size.

File: iOS/Flows/Conversation/ConversationSendOverlayView.swift [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+//  ConversationReplyOverlayView
+//
+//  Created by Martin Young on 10/13/21.
+//  Copyright © 2021 Benjamin Dodgson. All rights reserved.
+//
+
+import Foundation
+
+/// A view used by the conversation VC to designate where to drag and drop new messages.
+class ConversationSendOverlayView: View {
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/ConversationSendOverlayView.swift:12]: This is the dashed line box in the video.

File: iOS/Flows/Conversation/ConversationViewController.swift [hunk 5]
```diff
@@ -93,6 +76,24 @@ class ConversationViewController: FullScreenViewController,
...
+        self.conversationHeader.didSelect { [unowned self] in
+            self.didTapConversationTitle?()
+        }
+
+        self.contentContainer.addSubview(self.dateLabel)
+
+        self.messageInputAccessoryView.textView.$inputText.mainSink { _ in
+            guard let enabled = self.conversationController?.areTypingEventsEnabled, enabled else { return }
+            self.conversationController?.sendKeystrokeEvent(completion: nil)
+        }.store(in: &self.cancellables)
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/ConversationViewController.swift:96]: Just a little reorganization. No changes.

File: iOS/Flows/Conversation/ConversationViewController.swift [hunk 6]
```diff
@@ -107,12 +108,14 @@ class ConversationViewController: FullScreenViewController,
...
 
         self.collectionView.expandToSuperviewWidth()
-        let padding = self.conversationHeader.height + self.dateLabel.height + 100
-        self.collectionView.pin(.top, padding: padding)
+        let padding = self.dateLabel.height + 40
+        self.collectionView.match(.top, to: .bottom, of: self.conversationHeader, offset: padding)
         self.collectionView.height = self.view.height - padding
+
+        // Base the Y position of the date label on the top of the collection view.
+        self.dateLabel.match(.bottom, to: .top, of: self.collectionView, offset: -20)
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/ConversationViewController.swift:118]: Adjusting the layout to better account for the keyboard.

File: iOS/Flows/Conversation/ConversationViewController.swift [hunk 7]
```diff
@@ -281,6 +284,20 @@ class ConversationViewController: FullScreenViewController,
...
 
         self.collectionView.isUserInteractionEnabled = false
+
+        // Animate in the send overlay
+        self.contentContainer.addSubview(self.sendMessageOverlay)
+        self.sendMessageOverlay.alpha = 0
+        self.sendMessageOverlay.setState(nil)
+        UIView.animate(withDuration: Theme.animationDuration) {
+            self.sendMessageOverlay.alpha = 1
+        }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/ConversationViewController.swift:294]: Show the send overlay when the user starts dragging.

File: iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift [hunk 8]
```diff
@@ -303,7 +303,7 @@ class SwipeableInputAccessoryView: View, AttachmentViewControllerDelegate, UIGes
     private var initialPreviewOrigin: CGPoint?
     private var currentSendPosition: SendPosition?
     /// How far the preview view can be dragged left or right.
-    private let maxXOffset: CGFloat = 100
+    private let maxXOffset: CGFloat = 40
```
Comment (MartinJibber) [inline iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift:306]: Don't allow the message to slide off the screen.

File: iOS/Flows/Conversation/ConversationViewController.swift [hunk 9]
```diff
@@ -294,22 +311,25 @@ class ConversationViewController: FullScreenViewController,
         switch position {
         case .left, .middle:
             // Avoid animating content offset twice for redundant states
-            guard self.lastPreparedPosition.equalsOneOf(these: .left, .middle) else { break }
+            guard !self.lastPreparedPosition.equalsOneOf(these: .left, .middle) else { break }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/ConversationViewController.swift:314]: Whoops, logic error!

File: iOS/Flows/Conversation/ConversationViewController.swift [hunk 10]
```diff
@@ -321,12 +341,20 @@ class ConversationViewController: FullScreenViewController,
...
+        guard let currentIndexPath = self.collectionView.getCentermostVisibleIndex(),
+              let currentItem = self.dataSource.itemIdentifier(for: currentIndexPath),
+              case let .message(messageID) = currentItem else {
+
+                  // If there is no current message to reply to, assume we're sending a new message
+                  Task {
+                      await self.send(sendable)
+                  }
+                  return
+              }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/ConversationViewController.swift:354]: Accounting for a conversation with no messages.

File: Shared/Extensions/UI Extensions/UICollectionView+Extensions.swift (hunk 1 above)
Comment (Benjamindodgson) [reply Shared/Extensions/UI Extensions/UICollectionView+Extensions.swift:23]: Nice

File: iOS/Flows/Archives/Conversation/ConversationContentView.swift (hunk 2 above)
Comment (Benjamindodgson) [reply iOS/Flows/Archives/Conversation/ConversationContentView.swift:51]: lame sauce.

**Review by Benjamindodgson — APPROVED** [review]

**Pull request merged**